From 0e3f94ca0321daca728898b3f6273f55fa193b2d Mon Sep 17 00:00:00 2001
From: Tomas Hlavacek <tmshlvck@gmail.com>
Date: Tue, 5 May 2020 20:40:24 +0200
Subject: [PATCH] ARM: dts: mvebu: armada-385-turris-omnia: separate dts for
 SFP and PHY
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

The Turris Omnia board contains dual-personality ethernet NIC eth2 with
two operation modes: 1) SFP cage and 2) metalic 1000BASE-X PHY.
The differential pair carrying SGMII/1000BASE-X of eth2 is wired through
a multiplexor driven by the module-detect signal from the SFP cage.
The pin status can be read through I2C GPIO expander chip in userspace
when the sfp driver module is unloaded and / or in U-Boot prior to the
start of the kernel and the proper DTS file can be selected for the
(floolowing) boot.

Split DTS for Turris Omnia (that does not have any support for SFP cage)
into three files:
  armada-385-turris-omnia.dtsi - common base
  armada-385-turris-omnia-sfp.dts - DT with the SFP configuration and
PHY disabled
  armada-385-turris-omnia-phy.dts - DT with the PHY configuration and
SFP disabled

Current DSA driver does not allow multiple CPU ports and Turris Omnia
has two RGMII iterfaces wired between CPU and DSA switch.
Disable the second CPU port until there is a suitable driver to use it.

Signed-off-by: Tomas Hlavacek <tmshlvck@gmail.com>
Signed-off-by: Marek Behún <marek.behun@nic.cz>
---
 arch/arm/boot/dts/marvell/Makefile            |  3 ++-
 .../marvell/armada-385-turris-omnia-phy.dts   | 22 ++++++++++++++++++
 .../marvell/armada-385-turris-omnia-sfp.dts   | 23 +++++++++++++++++++
 ...omnia.dts => armada-385-turris-omnia.dtsi} |  2 --
 4 files changed, 47 insertions(+), 3 deletions(-)
 create mode 100644 arch/arm/boot/dts/marvell/armada-385-turris-omnia-phy.dts
 create mode 100644 arch/arm/boot/dts/marvell/armada-385-turris-omnia-sfp.dts
 rename arch/arm/boot/dts/marvell/{armada-385-turris-omnia.dts => armada-385-turris-omnia.dtsi} (99%)

diff --git a/arch/arm/boot/dts/marvell/Makefile b/arch/arm/boot/dts/marvell/Makefile
index 1e0f5ff492f7..958d26fc457d 100644
--- a/arch/arm/boot/dts/marvell/Makefile
+++ b/arch/arm/boot/dts/marvell/Makefile
@@ -44,7 +44,8 @@ dtb-$(CONFIG_MACH_ARMADA_38X) += \
 	armada-385-linksys-rango.dtb \
 	armada-385-linksys-shelby.dtb \
 	armada-385-synology-ds116.dtb \
-	armada-385-turris-omnia.dtb \
+	armada-385-turris-omnia-phy.dtb \
+	armada-385-turris-omnia-sfp.dtb \
 	armada-388-clearfog.dtb \
 	armada-388-clearfog-base.dtb \
 	armada-388-clearfog-pro.dtb \
diff --git a/arch/arm/boot/dts/marvell/armada-385-turris-omnia-phy.dts b/arch/arm/boot/dts/marvell/armada-385-turris-omnia-phy.dts
new file mode 100644
index 000000000000..706f6a2f8065
--- /dev/null
+++ b/arch/arm/boot/dts/marvell/armada-385-turris-omnia-phy.dts
@@ -0,0 +1,22 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Device Tree file for the Turris Omnia
+ *
+ * Copyright (C) 2016 Uwe Kleine-König <uwe@kleine-koenig.org>
+ * Copyright (C) 2016-2019 Tomas Hlavacek <tmshlvkc@gmail.com>
+ *
+ * Schematic available at https://www.turris.cz/doc/_media/rtrom01-schema.pdf
+ */
+
+/dts-v1/;
+
+#include "armada-385-turris-omnia.dtsi"
+
+&phy1 {
+	status = "okay";
+};
+
+&eth2 {
+	phy-mode = "sgmii";
+	phy = <&phy1>;
+};
diff --git a/arch/arm/boot/dts/marvell/armada-385-turris-omnia-sfp.dts b/arch/arm/boot/dts/marvell/armada-385-turris-omnia-sfp.dts
new file mode 100644
index 000000000000..b9f2b88834be
--- /dev/null
+++ b/arch/arm/boot/dts/marvell/armada-385-turris-omnia-sfp.dts
@@ -0,0 +1,23 @@
+// SPDX-License-Identifier: (GPL-2.0 OR MIT)
+/*
+ * Device Tree file for the Turris Omnia
+ *
+ * Copyright (C) 2016 Uwe Kleine-König <uwe@kleine-koenig.org>
+ * Copyright (C) 2016-2019 Tomas Hlavacek <tmshlvkc@gmail.com>
+ *
+ * Schematic available at https://www.turris.cz/doc/_media/rtrom01-schema.pdf
+ */
+
+/dts-v1/;
+
+#include "armada-385-turris-omnia.dtsi"
+
+&sfp {
+	status = "okay";
+};
+
+&eth2 {
+	phy-mode = "sgmii";
+	managed = "in-band-status";
+	sfp = <&sfp>;
+};
diff --git a/arch/arm/boot/dts/marvell/armada-385-turris-omnia.dts b/arch/arm/boot/dts/marvell/armada-385-turris-omnia.dtsi
similarity index 99%
rename from arch/arm/boot/dts/marvell/armada-385-turris-omnia.dts
rename to arch/arm/boot/dts/marvell/armada-385-turris-omnia.dtsi
index 2d8d319bec83..11be7bfa46f6 100644
--- a/arch/arm/boot/dts/marvell/armada-385-turris-omnia.dts
+++ b/arch/arm/boot/dts/marvell/armada-385-turris-omnia.dtsi
@@ -192,8 +192,6 @@ &eth2 {
 	 * handle and add managed = "in-band-status" property.
 	 */
 	status = "okay";
-	phy-mode = "sgmii";
-	phy-handle = <&phy1>;
 	phys = <&comphy5 2>;
 	sfp = <&sfp>;
 	buffer-manager = <&bm>;
-- 
2.42.0

